solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudSolrClient.java

```diff
@@ -1368,6 +1368,14 @@ private static LBHttpSolrClient createLBHttpSolrClient(Builder cloudSolrClientBu
      */
     @Deprecated
     public Builder() {}
+
+
+    /**Provide an already created {@link ClusterStateProvider} instance
+     *
+     */
+    public Builder(ClusterStateProvider stateProvider) {
+      this.stateProvider = stateProvider;
```

Comment:
Does it make more sense to just accept the `ZkStateReader` instead? And then have `build()` create the `ClusterStateProvider` where it checks the different scenarios (state reader vs zk hosts vs http client config).

Reply:
The Builder already accepts a ClusterStateProvider. So, this is OK

Reply:
That's fair. SG.